File: problems/cc/longest-palindrome.cc
// https://leetcode-cn.com/problems/longest-palindrome/

#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution
{
public:
    int longestPalindrome(string s)
    {
        unordered_map<char, int> chars;
        for (auto &&c : s)
        {
            ++chars[c];
        }
        int count = 0;
        for (auto &&item : chars)
        {
            if (item.second % 2)
            {
                ++count;
            }
        }
        return count == 0 ? s.size() : s.size() - count + 1;
    }
};

int main(int argc, char const *argv[])
{
    Solution s;
    cout << s.longestPalindrome("abccccdd") << endl;
    return 0;
}

File: problems/rs/longest-palindrome.rs
// https://leetcode-cn.com/problems/longest-palindrome/

use std::collections::HashMap;

#[derive(Debug)]
struct Solution {
}

impl Solution {
    pub fn longest_palindrome(s: String) -> i32 {
        let len = s.len();
        let mut chars: HashMap<char, i32> = HashMap::new();
        for c in s.chars() {
            if chars.contains_key(&c) {
                if let Some(n) = chars.get_mut(&c) {
                    *n += 1i32;
                }
            } else { chars.insert(c, 1i32); }
        }

        let mut count = 0;
        for val in chars.values() {
            if val % 2 == 1 { count += 1; } else { count += 0;}
        }

        if count == 0 { return len as i32; } else { return len as i32 - count + 1; }
    }
}

fn main() {
    println!("{:?}", Solution::longest_palindrome(String::from("abccccdd")));
}